src/realm/util/base64.cpp

```diff
@@ -0,0 +1,209 @@
+{
+    REALM_ASSERT_EX(input.size() < std::numeric_limits<size_t>::max() / 3, input.size());
+    size_t required_buffer_len = (input.size() * 3 + 3) / 4;
+    REALM_ASSERT_EX(out_buffer_len >= required_buffer_len, out_buffer_len, required_buffer_len);
+    static_cast<void>(out_buffer_len);
+    static_cast<void>(required_buffer_len);
+
+    // Guard against overlap (so that "restrict" works in the following)
+    REALM_ASSERT(input.data() + input.size() <= out_buffer
+                 || input.data() >= out_buffer + out_buffer_len);
+
+
+    const char* REALM_RESTRICT p = input.data();
+          char* REALM_RESTRICT o = out_buffer;
+
+    enum b64_byte_type {
+        equals     = 64,  // used as padding at the end
+        whitespace = 65,
+        invalid    = 66,
+    };
+
+    size_t bytes_written = 0;
+    size_t num_trailing_equals = 0;
+    unsigned int buffer = 0;
+    size_t i, buffer_size; // i is input sequence, buffer_size is only incremented for valid characters
+    for (i = 0, buffer_size = 0; i < input.size(); ++i) {
+        // classify the base64 character
+        unsigned int x = index_of_base64_byte(static_cast<unsigned char>(p[i]));
+        switch (x) {
+            case equals:     ++num_trailing_equals; continue;
+            case whitespace: continue; // ignore whitespace
+            case invalid:    return none;
+        }
+
+        if (num_trailing_equals > 0)
+            return none; // data after the end-padding
+
+        REALM_ASSERT_EX(x < 64, x);
+        buffer = buffer << 6 | x;
+        ++buffer_size;
+
+        if (buffer_size == 4) {
+            *o++ = (buffer >> 16) & 0xff;
+            *o++ = (buffer >>  8) & 0xff;
+            *o++ = (buffer >>  0) & 0xff;
+            buffer = 0;
+            buffer_size = 0;
+            bytes_written += 3;
+        }
+    }
+
+    // no padding
+    size_t extra = input.size() % 4;
+    if (num_trailing_equals == 0 && extra > 1) {
+        num_trailing_equals = 4 - extra;
+    }
+    
+    // trailing bytes
+    if (num_trailing_equals == 0) {
+        if (buffer_size != 0)
+            return none; // stuff was left in buffer, so input was not sufficiently padded.
+    }
+    else if (num_trailing_equals == 1) {
+        *o++ = (buffer >> 10) & 0xff;
+        *o++ = (buffer >>  2) & 0xff;
+        bytes_written += 2;
+    }
+    else if (num_trailing_equals == 2) {
+        *o++ = (buffer >>  4) & 0xff;
+        bytes_written += 1;
+    }
+    else {
+        return none;
+    }
+
+    return bytes_written;
+}
+
+Optional<std::vector<char>> base64_decode_to_vector(StringData encoded)
+{
+    size_t max_size = base64_decoded_size(encoded.size());
+    std::vector<char> decoded(max_size); // Throws
+    Optional<size_t> actual_size = base64_decode(encoded, decoded.data(), decoded.size());
+    if (!actual_size)
+        return none;
+
+    decoded.resize(*actual_size); // Throws
+    return decoded;
```

Comment:
In C++14, `return decoded;` _already_ treats `decoded` as an rvalue when performing overload resolution, so it'll be moved into the `util::Optional` that is constructed. There's no need to write `return std::move(decoded);`, and it will in fact inhibit optimizations if the type of `decoded` changes to match the return type in the future.